### PR TITLE
Blog Post: Announcing Dotty 0.6.0 and 0.7.0-RC1

### PR DESCRIPTION
**Pull request by allanrenucci**
@lampepfl/dotty-core Feel free to edit

### CODE REVIEW

File: docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+
+object Option {
+  def apply[T](x: T): Option[T] = if (x == null) None else Some(x)
+}
+```
+
+You can visit our website for more information about [enumerations](/docs/reference/enums/enums.html)
+and how we can use them to model [Algebraic Data Types](/docs/reference/enums/adts.html).
+
+### Unused Parameters [#3342](https://github.com/lampepfl/dotty/pull/3342)
```
Comment (allanrenucci) [inline docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: @nicolasstucki Do you want to do this?

**Comment by smarter** [comment]
This needs a section on improved GADT support and a section on improved completions in the IDE. I'm not available to write this myself thid week unfortunately.

File: docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md (hunk 1 above)
Comment (Blaisorblade) [reply docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: @nicolasstucki  Probably we should say a couple words about going from phantom types to unused parameters (https://github.com/lampepfl/dotty/issues/2040#issuecomment-369628306), probably in the release notes rather than in the docs for the latest release (to avoid confusing new users). Starting point:

> TL;DR: With unused parameters you can mark values of arbitrary types to be erased at runtime, instead of having to implement separate phantom types.

File: docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  }
+}
+
+object Option {
+  def apply[T](x: T): Option[T] = if (x == null) None else Some(x)
+}
+```
+
+You can visit our website for more information about [enumerations](/docs/reference/enums/enums.html)
+and how we can use them to model [Algebraic Data Types](/docs/reference/enums/adts.html).
```
[the remark this answers is not in the document]
Comment (allanrenucci) [reply docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: These links will become valid once the page is generated. An alternative would be to write absolute links (e.g. http://dotty.epfl.ch/docs/reference/enums/enums.html) so that they are also valid links in GitHub

**Comment by allanrenucci** [comment]
> This needs a section on improved GADT support and a section on improved completions in the IDE.

For reference, the PR which improved IDE completions is #3960. 

@smarter What improvements do you have in mind regarding GADT support? #3990?

**Comment by Blaisorblade** [comment]
@allanrenucci PRs on GADTs include #3918, #4034 (finding more...)

**Comment by smarter** [comment]
> What improvements do you have in mind regarding GADT support? #3990

Soundness was improved by https://github.com/lampepfl/dotty/pull/3918, although there are still holes related to variance until https://github.com/lampepfl/dotty/pull/4013 is merged. @liufengyun also did work related to gadt exhaustiveness checking

**Comment by Blaisorblade** [comment]
So #3645 was in fact fixed in December (so for last release), though it fits well under "ongoing GADT improvements". IIUC, even without #4013, invariant GADTs appear fixed, and with #4013 variant GADTs will be fixed as well.

File: docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (Blaisorblade) [reply docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: Sorry: I was asking if the docs in the target pages have been updated (and I thought not).

File: docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+<!--more-->
+
+This is our seventh scheduled release according to our [6-week release schedule](http://dotty.epfl.ch/docs/usage/version-numbers.html).
+The [previous technology preview](https://github.com/lampepfl/dotty/releases/tag/0.6.0-RC1) focussed
+on bug fixes and stability work.
+
+## What’s new in the 0.7.0-RC1 technology preview?
+
+### Enums Simplicification [#4003](https://github.com/lampepfl/dotty/pull/4003)
```
Comment (Duhemm) [inline docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: Typo: `Simplification`

File: docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+```
+
+Earlier Dotty releases had issues typechecking such interpreters.
+We have fixed multiple bugs about GADT type checking and exhaustiveness checking, especially for invariant GADTs, including
+[#3666](https://github.com/lampepfl/dotty/issues/3666),
+[#1754](https://github.com/lampepfl/dotty/issues/1754),
+[#3645](https://github.com/lampepfl/dotty/issues/3645),
+and improved handling of matches using repeated type variables
+[#4030](https://github.com/lampepfl/dotty/issues/4030).
+More test cases appear in [#3999](https://github.com/lampepfl/dotty/pull/3999).
```
Comment (allanrenucci) [inline docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: I would drop this line.

File: docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    case f: Fun[s, t]  => // Here T = s => t
+      (v: s) => eval(f.f(GenLit(v)))
+
+    case App(f, e)     => // Here f: Exp[s, T] and e: Exp[s]
+      eval(f)(eval(e))
+  }
+}
+```
+
+Earlier Dotty releases had issues typechecking such interpreters.
```
Comment (allanrenucci) [inline docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: I would say versions instead of releases.

Comment (smarter) [inline docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: I think it's important to note that Scala 2 also has unresolved issues typechecking GADTs, like with the neg example in https://github.com/lampepfl/dotty/pull/3990/files#diff-e7b3b26b928fa129081481ffc513af95

Comment (Blaisorblade) [reply docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: Added a bit; I think Dotty's already much better than Scalac, but I haven't tried enough personally to make strong claims this in release notes. Feel free to add stronger claims you feel comfortable with.

File: docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md [hunk 6]
```diff
@@ -162,7 +162,7 @@ object Interpreter {
 }
 ```
 
-Earlier Dotty releases had issues typechecking such interpreters.
+Earlier Scalac and Dotty releases had issues typechecking such interpreters.
```
Comment (smarter) [inline docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: This makes it sound like recent releases of scalac have fixed this, but they haven't. Also I would use the term "Scala 2" and not "Scalac".

Comment (smarter) [reply docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: The most convincing thing to do would be to give an example that doesn't typecheck in Dotty but does in Scala 2.

Comment (Blaisorblade) [reply docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: This program works in Scalac, while trickier ones trigger bugs in Scalac or Dotty. I propose to write "Scala 2 and Dotty have issues [...]", since we fixed some.

```scala
case class MapS[S, T](xs: Exp[List[S]], f: Exp[S => T]) extends Exp[List[T]]
case class Map[S, T, That](xs: Exp[List[S]], f: Exp[S => T])(implicit val cbf: CanBuildFrom[List[S], T, That]) extends Exp[That]

//...

    case m @ Map(xs, f)    =>
      //val f1: Exp[Any => Any] = f //accepted by Scalac, rejected by Dotty!
      eval(xs).map(eval(f))(m.cbf)

    case MapS(xs, f) =>
      eval(xs) map eval(f) //rejected by Dotty, not 100% sure why. Passing the 

   //alternatively:
    case m: MapS[s, t] =>
      //(eval(m.xs) map eval(m.f))(List.canBuildFrom) //fails in Dotty
      (eval(m.xs) map eval(m.f))(List.canBuildFrom[t]) //succeeds in Dotty

```

Failures in https://github.com/lampepfl/dotty/pull/4069 might be relevant.

File: docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md [hunk 7]
```diff
@@ -159,8 +159,8 @@ object Interpreter {
 }
 ```
 
-Earlier Scalac and Dotty releases had issues typechecking such interpreters.
-We fixed multiple bugs about GADT type checking and exhaustiveness checking, including
+Scala 2 and Dotty have issues typechecking such interpreters.
+In this release we fixed multiple bugs about GADT type checking and exhaustiveness checking, including
```
Comment (Blaisorblade) [inline docs/blog/_posts/2018-03-05-seventh-dotty-milestone-release.md]: @smarter @allanrenucci WDYT? A bit less bold which seems appropriate.

**Comment by smarter** [comment]
I've written an alternative version of the GADT section, I'll let others decide which one is better.
